diesel/src/expression/cast.rs

```diff
@@ -0,0 +1,152 @@
+//! SQL `CAST(expr AS sql_type)` expression support
+
+use crate::expression::{AppearsOnTable, Expression, SelectableExpression, ValidGrouping};
+use crate::query_source::aliasing::{AliasSource, FieldAliasMapper};
+use crate::result::QueryResult;
+use crate::{query_builder, query_source, sql_types};
+
+use std::marker::PhantomData;
+
+pub(crate) mod private {
+    use super::*;
+
+    #[derive(Debug, Clone, Copy, diesel::query_builder::QueryId, sql_types::DieselNumericOps)]
+    pub struct Cast<E, ST> {
+        pub(super) expr: E,
+        pub(super) sql_type: PhantomData<ST>,
+    }
+}
+pub(crate) use private::Cast;
+
+impl<E, ST> Cast<E, ST> {
+    pub(crate) fn new(expr: E) -> Self {
+        Self {
+            expr,
+            sql_type: PhantomData,
+        }
+    }
+}
+
+impl<E, ST, GroupByClause> ValidGrouping<GroupByClause> for Cast<E, ST>
+where
+    E: ValidGrouping<GroupByClause>,
+{
+    type IsAggregate = E::IsAggregate;
+}
+
+impl<E, ST, QS> SelectableExpression<QS> for Cast<E, ST>
+where
+    Cast<E, ST>: AppearsOnTable<QS>,
+    E: SelectableExpression<QS>,
+{
+}
+
+impl<E, ST, QS> AppearsOnTable<QS> for Cast<E, ST>
+where
+    Cast<E, ST>: Expression,
+    E: AppearsOnTable<QS>,
+{
+}
+
+impl<E, ST> Expression for Cast<E, ST>
+where
+    E: Expression,
+    ST: sql_types::SingleValue,
+{
+    type SqlType = ST;
+}
+impl<E, ST, DB> query_builder::QueryFragment<DB> for Cast<E, ST>
+where
+    E: query_builder::QueryFragment<DB>,
+    DB: diesel::backend::Backend,
+    ST: KnownCastSqlTypeName<DB>,
+{
+    fn walk_ast<'b>(&'b self, mut out: query_builder::AstPass<'_, 'b, DB>) -> QueryResult<()> {
+        out.push_sql("CAST(");
+        self.expr.walk_ast(out.reborrow())?;
+        out.push_sql(" AS ");
+        out.push_sql(ST::sql_type_name());
+        out.push_sql(")");
+        Ok(())
+    }
+}
+
+/// We know what to write as `sql_type` in the `CAST(expr AS sql_type)` SQL for
+/// `Self`
+///
+/// That is what is returned by `Self::sql_type_name()`
+#[diagnostic::on_unimplemented(
+    note = "In order to use `CAST`, it is necessary that Diesel knows how to express the name \
+		of this type in the given backend.",
+    note = "This can be PRed into Diesel if the type is a standard SQL type."
```

Comment:
We likely want to change this message to something more active like: If you run into this error message and believe that this cast should be supported open a PR that adds that trait implementation there (with link to the source or file path)